warp_drive/training/lightning-trainer.py

```diff
@@ -0,0 +1,520 @@
+            policy
+            for policy in self.policies
+            if self._get_config(["policy", policy, "to_train"])
+        ]
+
+        # Flag indicating whether there needs to be separate placeholders / arrays
+        # for observation, actions and rewards, for each policy
+        self.create_separate_placeholders_for_each_policy = (
+            create_separate_placeholders_for_each_policy
+        )
+        # Note: separate placeholders are needed only when there are
+        # multiple policies
+        if self.create_separate_placeholders_for_each_policy:
+            assert len(self.policies) > 1
+
+        # Number of iterations algebra
+        self.num_episodes = self._get_config(["trainer", "num_episodes"])
+        self.training_batch_size = self._get_config(["trainer", "train_batch_size"])
+        self.num_envs = self._get_config(["trainer", "num_envs"])
+
+        self.training_batch_size_per_env = self.training_batch_size // self.num_envs
+        assert self.training_batch_size_per_env > 0
+
+        self.block = self.cuda_envs.cuda_function_manager.block
+        self.data_manager = self.cuda_envs.cuda_data_manager
+        self.function_manager = self.cuda_envs.cuda_function_manager
+
+        # Seeding
+        seed = self.config["trainer"].get("seed", np.int32(time.time()))
+        self.cuda_sample_controller = CUDASampler(self.cuda_envs.cuda_function_manager)
+        self.cuda_sample_controller.init_random(seed)
+
+        torch.manual_seed(seed)
+        random.seed(seed)
+        np.random.seed(seed)
+
+        # Define models, optimizers, and learning rate schedules
+        self.models = {}
+        self.optimizers = {}
+        self.lr_schedules = {}
+
+        # For logging episodic reward
+        self.num_completed_episodes = {}
+        self.episodic_reward_sum = {}
+        self.reward_running_sum = {}
+
+        # Indicates the current timestep of the policy model
+        self.current_timestep = {}
+
+        self.total_steps = self.cuda_envs.episode_length * self.num_episodes
+        self.num_iters = self.total_steps // self.training_batch_size
+
+        for policy in self.policies:
+            self.current_timestep[policy] = 0
+            policy_model_config = self._get_config(["policy", policy, "model"])
+            if policy_model_config["type"] == "fully_connected":
+                self.models[policy] = FullyConnected(
+                    self.cuda_envs,
+                    policy_model_config["fc_dims"],
+                    policy,
+                    self.policy_tag_to_agent_id_map,
+                    create_separate_placeholders_for_each_policy,
+                    obs_dim_corresponding_to_num_agents,
+                )
+            else:
+                raise NotImplementedError
+
+        # Load the model parameters (if model checkpoints are specified)
+        # Note: Loading the model checkpoint may also update the current timestep!
+        self.load_model_checkpoint()
```

Comment:
Not sure if this is required.

Reply:
This is to initialize the NN model with any user-specified training checkpoint.